src/storage/txn/actions/acquire_pessimistic_lock.rs

```diff
@@ -82,148 +85,58 @@ pub fn acquire_pessimistic_lock<S: Snapshot>(
-                            // to care about it.
-                            let conflict_info = ConflictInfo {
-                                conflict_start_ts: write.start_ts,
-                                conflict_commit_ts: commit_ts,
-                            };
-                            return Err(conflict_info.into_write_conflict_error(
-                                reader.start_ts,
-                                primary.to_vec(),
-                                key.to_raw()?,
-                            ));
-                        }
-                        Err(e)
-                    },
-                )?;
-
-                if need_load_value {
-                    val = Some(reader.load_data(&key, write)?);
-                } else if need_check_existence {
-                    val = Some(vec![]);
-                }
+    let current_lock = reader.load_lock(&key)?;
+    if let Some(lock) = current_lock.as_ref() {
+        if !is_shared_lock_req {
+            // handle exclusive lock.
+            if lock.is_shared() {
+                return Err(ErrorInner::KeyIsLocked(
+                    current_lock.unwrap().into_lock_info(key.into_raw()?),
+                )
+                .into());
             }
-        }
-        // Previous write is not loaded.
-        let (prev_write_loaded, prev_write) = (false, None);
-        let old_value = load_old_value(
-            need_old_value,
-            need_load_value,
-            val.as_ref(),
-            reader,
-            &key,
-            for_update_ts,
-            prev_write_loaded,
-            prev_write,
-        )?;
-
-        // Overwrite the lock with small for_update_ts
-        if for_update_ts > lock.for_update_ts {
-            let lock = PessimisticLock {
-                primary: primary.into(),
-                start_ts: reader.start_ts,
-                ttl: lock_ttl,
+            return handle_existing_exclusive_lock(
+                txn,
+                reader,
+                key,
+                primary,
+                should_not_exist,
+                lock_ttl,
                 for_update_ts,
+                need_value,
+                need_check_existence,
                 min_commit_ts,
-                last_change: lock.last_change.clone(),
-                is_locked_with_conflict: lock.is_pessimistic_lock_with_conflict(),
-            };
-            txn.put_pessimistic_lock(key, lock, false);
+                need_old_value,
+                allow_lock_with_conflict,
+                current_lock.unwrap(),
+            );
         } else {
-            MVCC_DUPLICATE_CMD_COUNTER_VEC
-                .acquire_pessimistic_lock
-                .inc();
+            // handle shared lock.
+            if !lock.is_shared() {
+                return Err(ErrorInner::KeyIsLocked(
+                    current_lock.unwrap().into_lock_info(key.into_raw()?),
+                )
+                .into());
+            }
```

Comment:
So we will reject a shared lock req when there is an existing exclusive pessimistic lock even they come from the same txn (with same ts)? I see the [previous impl](https://github.com/tikv/tikv/blob/e31634578b89a10450a023899d699c6ea4e8c33b/src/storage/txn/actions/acquire_pessimistic_lock.rs#L115-L223) allows this case.

Reply:
Reject the exclusive lock downgrade is more strict. Since we haven't well design the desired behavior, I think it's ok to return error here.

The client of TiKV may not send shared lock request when adding shared lock above an exclusive lock (and prewrite with `Op::Lock` as well).